mos-platform/cx16/cx16.h

```diff
@@ -296,5 +305,164 @@ struct __emul {
 /* An array window into the half Mebibyte or two Mebibytes of banked RAM */
 #define BANK_RAM        ((volatile unsigned char *)0xA000)
 
-/* End of cX16.h */
-#endif
+/*****************************************************************************/
+/*                          CX16 kernal functions                            */
+/*****************************************************************************/
+
+/* Kernel function structure type definitions */
+typedef struct
+{
+    unsigned char year, mon, day, hour, min, sec, jif;
+} cx16_date_time_t; /* cx16_k_clock_get_date_time() */
+
+typedef struct
+{
+    unsigned int width, height;
+    unsigned char depth;
+} cx16_fb_info_t; /* cx16_k_fb_get_info_depth() */
+
+typedef struct
+{
+    void *fb_init;
+    void *fb_get_info;
+    void *fb_set_palette;
+    void *fb_cursor_position;
+    void *fb_cursor_next_line;
+    void *fb_get_pixel;
+    void *fb_get_pixels;
+    void *fb_set_pixel;
+    void *fb_set_pixels;
+    void *fb_set_8_pixels;
+    void *fb_set_8_pixels_opaque;
+    void *fb_fill_pixels;
+    void *fb_filter_pixels;
+    void *fb_move_pixels;
+} graph_fb_functions_t; /* cx16_k_graph_init() */
+
+typedef struct
+{
+    int x, y;
+} graph_pos_t;  /* fb_graph_put_char() */
+
+typedef struct
+{
+    int x, y;
+} mouse_pos_t; /* cx16_k_mouse_get() */
+
+typedef struct
+{
+    unsigned char mode, columns, rows;
+} screen_mode_info_t;   /* cx16_k_screen_mode_get() */
+
+/* Kernal-level functions */
+void cx16_k_clock_get_date_time(cx16_date_time_t *datetime_ptr);
```

Comment:
All of these that are implemented in assembly and cannot call back into the user program or anything in the platform libraries should be marked as `__attribute__((leaf))` (i.e., they're leaf calls from the POV of the compiler). Otherwise, the compiler has no idea what they might call, so it will have to assume anything that calls them can recurse, which forces use of a cc65-style soft stack.

It should be that all or almost all of these are leaves, but that's something to quickly verify for each wrapper.

Reply:
As discussed, doing these leaf changes in later PR (since they also affect cbm etc.).